File: src/server/google.ts
/**
 * ตรวจ Google ID token
 *
 * ใช้รูปแบบที่ client ฝั่งไหนก็ทำได้เหมือนกัน — เว็บ, iOS, Android ต่างขอ ID token
 * จาก Google SDK ของตัวเอง แล้วส่งมาให้ server ตรวจ ไม่ต้องมี redirect flow
 * ซึ่งจำเป็นสำหรับ crossplay เพราะแอปมือถือทำ redirect แบบเว็บไม่ได้สะดวก
 *
 * ตรวจลายเซ็นเองด้วยกุญแจสาธารณะของ Google แทนที่จะเรียก tokeninfo ทุกครั้ง
 * จะได้ไม่ต้องยิงออกนอกทุกการล็อกอิน
 */

import { createPublicKey, createVerify } from "node:crypto";

export const GOOGLE_JWKS_URL = "https://www.googleapis.com/oauth2/v3/certs";
const VALID_ISSUERS = new Set(["accounts.google.com", "https://accounts.google.com"]);
const JWKS_TTL_MS = 60 * 60 * 1000;

export interface GoogleIdentity {
  sub: string;
  email: string | null;
  emailVerified: boolean;
  name: string | null;
}

export interface GoogleVerifierOptions {
  /** client id ที่ยอมรับ — ปกติมีหลายตัวเพราะเว็บกับมือถือใช้คนละตัว */
  clientIds: string[];
  jwksUrl?: string;
  fetcher?: typeof fetch;
  now?: () => number;
}

interface Jwk {
  kid: string;
  kty: string;
  alg?: string;
  use?: string;
  n: string;
  e: string;
}

export class GoogleVerifier {
  private readonly clientIds: Set<string>;
  private readonly jwksUrl: string;
  private readonly fetcher: typeof fetch;
  private readonly now: () => number;
  private cache: { keys: Map<string, Jwk>; expiresAt: number } | null = null;

  constructor(options: GoogleVerifierOptions) {
    this.clientIds = new Set(options.clientIds.filter(Boolean));
    this.jwksUrl = options.jwksUrl ?? GOOGLE_JWKS_URL;
    this.fetcher = options.fetcher ?? fetch;
    this.now = options.now ?? Date.now;
  }

  get enabled(): boolean {
    return this.clientIds.size > 0;
  }

  /** คืน identity เมื่อ token ถูกต้องทุกประการ ไม่งั้นคืน null */
  async verify(idToken: unknown): Promise<GoogleIdentity | null> {
    if (!this.enabled || typeof idToken !== "string") return null;

    const parts = idToken.split(".");
    if (parts.length !== 3) return null;
    const [rawHeader, rawPayload, rawSignature] = parts;

    let header: { kid?: string; alg?: string };
    let payload: Record<string, unknown>;
    try {
      header = JSON.parse(Buffer.from(rawHeader, "base64url").toString());
      payload = JSON.parse(Buffer.from(rawPayload, "base64url").toString());
    } catch {
      return null;
    }

    // รับเฉพาะ RS256 — ป้องกัน alg confusion เช่นการยัด alg:none เข้ามา
    if (header.alg !== "RS256" || !header.kid) return null;

    const jwk = await this.findKey(header.kid);
    if (!jwk) return null;

    let verified = false;
    try {
      const key = createPublicKey({ key: jwk as unknown as JsonWebKey, format: "jwk" });
      verified = createVerify("RSA-SHA256")
        .update(`${rawHeader}.${rawPayload}`)
        .verify(key, Buffer.from(rawSignature, "base64url"));
    } catch {
      return null;
    }
    if (!verified) return null;

    if (typeof payload.iss !== "string" || !VALID_ISSUERS.has(payload.iss)) return null;
    if (typeof payload.aud !== "string" || !this.clientIds.has(payload.aud)) return null;
    if (typeof payload.sub !== "string" || payload.sub.length === 0) return null;

    const nowSeconds = Math.floor(this.now() / 1000);
    if (typeof payload.exp !== "number" || payload.exp <= nowSeconds) return null;
    // เผื่อนาฬิกาคลาดกันเล็กน้อยระหว่างเครื่อง
    if (typeof payload.iat === "number" && payload.iat > nowSeconds + 300) return null;

    return {
      sub: payload.sub,
      email: typeof payload.email === "string" ? payload.email.toLowerCase() : null,
      emailVerified: payload.email_verified === true || payload.email_verified === "true",
      name: typeof payload.name === "string" ? payload.name : null,
    };
  }

  private async findKey(kid: string): Promise<Jwk | null> {
    const cached = this.cache;
    if (cached && cached.expiresAt > this.now()) {
      const key = cached.keys.get(kid);
      if (key) return key;
    }
    // ไม่รู้จัก kid นี้ อาจเพราะ Google หมุนกุญแจ ลองโหลดใหม่หนึ่งครั้ง
    await this.refresh();
    return this.cache?.keys.get(kid) ?? null;
  }

  private async refresh(): Promise<void> {
    try {
      const response = await this.fetcher(this.jwksUrl);
      if (!response.ok) return;
      const body = (await response.json()) as { keys?: Jwk[] };
      const keys = new Map<string, Jwk>();
      for (const key of body.keys ?? []) {
        if (key.kty === "RSA" && key.kid) keys.set(key.kid, key);
      }
      if (keys.size > 0) this.cache = { keys, expiresAt: this.now() + JWKS_TTL_MS };
    } catch {
      // โหลดกุญแจไม่ได้ก็ปล่อยให้ verify คืน null ดีกว่าปล่อยผ่าน
    }
  }
}

File: apps/api/src/makthai/services/google.py
"""ล็อกอินด้วย Google

ใช้รูปแบบที่ client ฝั่งไหนก็ทำได้เหมือนกัน — เว็บ ไอโอเอส แอนดรอยด์ ต่างขอ ID token
จาก SDK ของแพลตฟอร์มตัวเอง แล้วส่งมาให้เซิร์ฟเวอร์ตรวจ ไม่ต้องมี redirect flow
ซึ่งจำเป็นสำหรับแอปมือถือ

ตรวจลายเซ็นเองด้วยกุญแจสาธารณะของ Google แทนการเรียก tokeninfo ทุกครั้ง
จะได้ไม่ต้องยิงออกนอกทุกการล็อกอิน
"""

from __future__ import annotations

import base64
import json
import time
from dataclasses import dataclass
from typing import Any, cast

import httpx
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding, rsa

GOOGLE_JWKS_URL = "https://www.googleapis.com/oauth2/v3/certs"
VALID_ISSUERS = frozenset({"accounts.google.com", "https://accounts.google.com"})
JWKS_TTL = 3600.0
#: เผื่อนาฬิกาคลาดกันเล็กน้อยระหว่างเครื่อง
CLOCK_SKEW = 300


@dataclass(frozen=True, slots=True)
class GoogleIdentity:
    sub: str
    email: str | None
    email_verified: bool
    name: str | None


def _b64(value: str) -> bytes:
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))


def _int(value: str) -> int:
    return int.from_bytes(_b64(value), "big")


class GoogleVerifier:
    def __init__(
        self,
        client_ids: list[str],
        *,
        jwks_url: str = GOOGLE_JWKS_URL,
        fetch: Any = None,
        now: Any = time.time,
    ) -> None:
        self.client_ids = {value for value in client_ids if value}
        self.jwks_url = jwks_url
        self._fetch = fetch
        self._now = now
        self._keys: dict[str, dict[str, str]] = {}
        self._expires_at = 0.0

    @property
    def enabled(self) -> bool:
        return bool(self.client_ids)

    async def verify(self, id_token: object) -> GoogleIdentity | None:
        """คืนตัวตนเมื่อโทเคนถูกต้องทุกประการ ไม่งั้นคืน None"""
        if not self.enabled or not isinstance(id_token, str):
            return None
        parts = id_token.split(".")
        if len(parts) != 3:
            return None
        raw_header, raw_payload, raw_signature = parts

        try:
            header = json.loads(_b64(raw_header))
            payload = json.loads(_b64(raw_payload))
        except (ValueError, json.JSONDecodeError):
            return None

        # รับเฉพาะ RS256 ป้องกัน alg confusion เช่นการยัด alg none เข้ามา
        if header.get("alg") != "RS256" or not header.get("kid"):
            return None

        jwk = await self._key(str(header["kid"]))
        if jwk is None:
            return None
        if not self._signature_ok(jwk, f"{raw_header}.{raw_payload}", _b64(raw_signature)):
            return None

        if payload.get("iss") not in VALID_ISSUERS:
            return None
        if payload.get("aud") not in self.client_ids:
            return None
        sub = payload.get("sub")
        if not isinstance(sub, str) or not sub:
            return None

        now = int(self._now())
        exp = payload.get("exp")
        if not isinstance(exp, int | float) or exp <= now:
            return None
        iat = payload.get("iat")
        if isinstance(iat, int | float) and iat > now + CLOCK_SKEW:
            return None

        email = payload.get("email")
        return GoogleIdentity(
            sub=sub,
            email=email.lower() if isinstance(email, str) else None,
            email_verified=payload.get("email_verified") in (True, "true"),
            name=payload.get("name") if isinstance(payload.get("name"), str) else None,
        )

    def _signature_ok(self, jwk: dict[str, str], signed: str, signature: bytes) -> bool:
        try:
            numbers = rsa.RSAPublicNumbers(e=_int(jwk["e"]), n=_int(jwk["n"]))
            key = numbers.public_key()
            key.verify(signature, signed.encode(), padding.PKCS1v15(), hashes.SHA256())
        except Exception:
            return False
        return True

    async def _key(self, kid: str) -> dict[str, str] | None:
        if self._expires_at > self._now() and kid in self._keys:
            return self._keys[kid]
        # ไม่รู้จักกุญแจนี้ อาจเพราะ Google หมุนกุญแจ ลองโหลดใหม่หนึ่งครั้ง
        await self._refresh()
        return self._keys.get(kid)

    async def _refresh(self) -> None:
        try:
            body = await self._load()
        except Exception:
            # โหลดกุญแจไม่ได้ก็ปล่อยให้ verify คืน None ดีกว่าปล่อยผ่าน
            return
        keys = {
            key["kid"]: key
            for key in body.get("keys", [])
            if key.get("kty") == "RSA" and key.get("kid")
        }
        if keys:
            self._keys = keys
            self._expires_at = self._now() + JWKS_TTL

    async def _load(self) -> dict[str, Any]:
        if self._fetch is not None:
            body = await self._fetch(self.jwks_url)
            return cast(dict[str, Any], body)
        async with httpx.AsyncClient(timeout=10) as client:
            response = await client.get(self.jwks_url)
            response.raise_for_status()
            return cast(dict[str, Any], response.json())


__all__ = ["GOOGLE_JWKS_URL", "GoogleIdentity", "GoogleVerifier"]
